shared/src/types/vote_extensions/validator_set_update.rs

```diff
@@ -0,0 +1,297 @@
+        Ok(EthAddr(ethereum::H160(inner)))
+    }
+}
+
+impl BorshSchema for EthAddr {
+    fn add_definitions_recursively(
+        definitions: &mut std::collections::HashMap<
+            borsh::schema::Declaration,
+            borsh::schema::Definition,
+        >,
+    ) {
+        let fields =
+            borsh::schema::Fields::UnnamedFields(borsh::maybestd::vec![
+                <[u8; 20]>::declaration()
+            ]);
+        let definition = borsh::schema::Definition::Struct { fields };
+        Self::add_definition(Self::declaration(), definition, definitions);
+    }
+
+    fn declaration() -> borsh::schema::Declaration {
+        "validator_set_update::EthAddr".into()
+    }
+}
+
+/// Contains the digest of all signatures from a quorum of
+/// validators for a [`Vext`].
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, BorshSchema)]
+pub struct VextDigest {
+    /// A mapping from a validator address to a [`Signature`].
+    pub signatures: HashMap<Address, Signature>,
+    /// The addresses of the validators in the new [`Epoch`],
+    /// and their respective voting power.
+    pub voting_powers: VotingPowersMap,
+}
+
+impl VextDigest {
+    /// Decompresses a set of signed [`Vext`] instances.
+    pub fn decompress(self, epoch: Epoch) -> Vec<SignedVext> {
+        let VextDigest {
+            signatures,
+            voting_powers,
+        } = self;
+
+        let mut extensions = vec![];
+
+        for (validator_addr, signature) in signatures.into_iter() {
+            let voting_powers = voting_powers.clone();
+            let data = Vext {
+                validator_addr,
+                voting_powers,
+                epoch,
+            };
+            extensions.push(SignedVext::new_from(data, signature));
+        }
+        extensions
+    }
+
+    /// Returns an Ethereum ABI encoded string with the
+    /// params to feed to the Ethereum bridge smart contracts.
+    pub fn abi_params(&self) -> String {
+        todo!()
+    }
+}
+
+/// Represents a [`Vext`] signed by some validator, with
+/// an Ethereum key.
+pub type SignedVext = Signed<Vext, SerializeWithAbiEncode>;
+
+impl SignedVext {
+    /// Serialize a [`Vext`] to be signed.
+    fn serialize_vext(ext: &Vext) -> [u8; 32] {
+        AbiEncode::signed_keccak256(&[
+            Token::String("updateValidatorsSet".into()),
+            Token::FixedBytes(
+                ext.voting_powers.get_bridge_hash(ext.epoch).to_vec(),
+            ),
+            Token::FixedBytes(
+                ext.voting_powers.get_governance_hash(ext.epoch).to_vec(),
+            ),
+            epoch_to_token(ext.epoch),
+        ])
+    }
+
+    /// Sign this [`Vext`] with an Ethereum key.
+    ///
+    /// For more information, check the Ethereum bridge smart contract code:
+    ///   - <https://github.com/anoma/ethereum-bridge/blob/main/contracts/contract/Bridge.sol#L186>
+    pub fn new_abi_encoded(keypair: &common::SecretKey, ext: Vext) -> Self {
```

Comment:
Can't we just implement a `new` method in the `types/proto.rs` defns for `Signed<T,S>`? And add an ABI marker struct there. That seemed like the whole point of that change and we get type API consistency.